server/src/main/java/io/crate/planner/operators/AbstractJoinPlan.java

```diff
@@ -69,13 +86,94 @@ public LogicalPlan rhs() {
 
     @Override
     public List<Symbol> outputs() {
-        if (joinType == JoinType.SEMI) {
-            return lhs.outputs();
-        } else {
-            return Lists.concat(lhs.outputs(), rhs.outputs());
+        return outputs;
+    }
+
+    @Nullable
+    protected AbstractJoinPlan.PrunedOutputsResult pruneOutputs(SequencedCollection<Symbol> outputsToKeep) {
```

Comment:
This adapts the pruning logic to also deal with outputs and make it usable for all join subclasses.